src/block/block_cache.c

```diff
@@ -442,90 +447,91 @@ __wt_blkcache_put(WT_SESSION_IMPL *session, wt_off_t offset, size_t size, uint32
         WT_STAT_CONN_INCR(session, block_cache_bypass_overhead_put);
-        return (WT_BLKCACHE_BYPASS);
+        return (0);
     }
-    /*
-     * Allocate space in the cache outside of the critical section. In the unlikely event that we
-     * fail to allocate metadata, or if the item exists and the caller did not check for that prior
-     * to calling this function, we will free the space.
-     */
-    WT_RET(__blkcache_alloc(session, size, &data_ptr));
 
     /*
-     * We will use the item id structure as a byte array for (1) computing a hash, and (2) for
-     * comparing with the items in the hash table.
+     * Allocate and initialize space in the cache outside of the critical section. In the unlikely
+     * event that we fail an allocation, free the space. NVRAM allocations can fail if there's no
+     * available memory, treat it as a cache-full failure.
      */
-    id.checksum = checksum;
-    id.offset = offset;
-    id.size = (uint32_t)size;
-    id.fid = S2BT(session)->id;
-    hash = __wt_hash_city64(&id, sizeof(id));
-
+    WT_RET(__blkcache_alloc(session, data->size, &data_ptr));
+    if (data_ptr == NULL)
+        return (0);
+    WT_ERR(__wt_calloc_one(session, &blkcache_store));
+    blkcache_store->fid = S2BT(session)->id;
+    WT_ASSERT(session, addr_size <= sizeof(blkcache_store->addr));
+    blkcache_store->addr_size = (uint8_t)addr_size;
+    memcpy(blkcache_store->addr, addr, addr_size);
+    blkcache_store->data = data_ptr;
+    blkcache_store->data_size = WT_STORE_SIZE(data->size);
+    memcpy(blkcache_store->data, data->data, data->size);
+
+    hash = __wt_hash_city64(addr, addr_size);
     bucket = hash % blkcache->hash_size;
     __wt_spin_lock(session, &blkcache->hash_locks[bucket]);
-    TAILQ_FOREACH (blkcache_item, &blkcache->hash[bucket], hashq) {
-        if (blkcache_item->id.checksum == id.checksum && blkcache_item->id.offset == id.offset &&
-          blkcache_item->id.size == id.size && blkcache_item->id.fid == id.fid)
-            goto item_exists;
-    }
 
-    WT_ERR(__wt_calloc_one(session, &blkcache_item));
-    blkcache_item->id = id;
-    blkcache_item->data = data_ptr;
+    /*
+     * In the case of a read, check if the block is already in the cache: it's possible because two
+     * readers can attempt to cache the same overflow block because overflow blocks aren't cached at
+     * the btree level. Collisions are relatively unlikely because other page types are cached at
+     * higher levels and reads of those tree pages are single-threaded so the page can be converted
+     * to its in-memory form before reader access. In summary, because collisions are unlikely, the
+     * allocation and copying remains outside of the bucket lock and collision check. Writing a
+     * block is single-threaded at a higher level, and as there should never be a collision, only
+     * check in diagnostic mode.
+     */
+#if !defined(HAVE_DIAGNOSTIC)
```

Comment:
I did see blocks being updated for metadata files, remember we had that conversation? So until we figure out how to weed out of those files, we need to make the check not only in the diagnostic mode.

Reply:
> I did see blocks being updated for metadata files, remember we had that conversation? So until we figure out how to weed out of those files, we need to make the check not only in the diagnostic mode, and we don't assert if we find a block being updated on a write.

Do you have any test that could reproduce that failure?

My guess is you saw an overflow item (overflow items are more common in metadata files than they are in data files, checkpoints, for example, can get quite large). I've now run a bunch of block-cache workloads, and overflow blocks are the only block types I've ever seen be a problem. Absent a reproducer or some explanation of why another type of block could be a problem, I'm inclined to wait for something to fire.

Reply:
I should be able to reproduce. We can approve this pull request and open another ticket to try and reproduce. I won't get to this until at least next week, so no need to hold this pull request.